Output dim tag of torch.full

In #113, it shows that the dim tag of the output shape of `torch.full` is not set correctly. An example to reproduce the issue looks like this:
```
def test_full_out_dim_tag():
  n_batch, n_feat = 3, 5

  def model_func(wrapped_import, inputs: torch.Tensor):
    if typing.TYPE_CHECKING or not wrapped_import:
      import torch
    else:
      torch = wrapped_import("torch")
    full = torch.full((inputs.shape[0],), 42)
    arange = torch.arange(full .shape[0])
    return arange + full 

  rnd = numpy.random.RandomState(42)
  x = rnd.normal(0., 1., (n_batch, n_feat)).astype("float32")
  verify_torch_and_convert_to_returnn(model_func, inputs=x)
```
which fails in `pytorch_to_returnn/torch/nn/modules/operator.py`, line 594, in `_unify_tensor_axes_returnn_meta` with
```
    line: assert all(dim.dimension == d.dimension for d in dims_for_axis if d is not None and d.dimension != 1), (
            f"invalid input {x} axis {i} dim {dim}")
    locals:
      all = <builtin> <built-in function all>
      dim = <local> Dim{'Range:range'(3)}
      dim.dimension = <local> 3
      d = <not found>
      d.dimension = <not found>
      dims_for_axis = <local> [Dim{'Range:range'(3)}, Dim{B}]
AssertionError: invalid input <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'Cast_output' [F|'Range:range'(3)] axes id> axis 0 dim Dim{'Range:range'(3)}
```
(I'll add a full stack trace and log later).

We can see that the dim tag in the output shape of `torch.full` is wrong, but in `returnn_data` it's still correct. When creating a range based on that, the wrong dim tag of the output shape is also used on RETURNN side, which will lead to the error when trying to add both tensors.
```
      arange = <local> <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'Range_output' [F|'Range:range'(3)] axes id>
      full = <local> <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'FullStatic_const' [B] axes id>
```

--- a/pytorch_to_returnn/naming/tensor.py
+++ b/pytorch_to_returnn/naming/tensor.py
@@ -19,7 +19,7 @@ class TensorEntry:
   validated_to_torch_tf_feed_dict: Optional[Dict[tf.Tensor, numpy.ndarray]] = None
   validated_to_torch_tf_sizes_feed_dict: Optional[Dict[tf.Tensor, numpy.ndarray]] = None
   is_param: bool = False
-  is_const: bool = False  # e.g. via from_numpy, empty, zeros, etc
+  is_const: bool = False  # e.g. via from_numpy, empty, zeros, etc (without dependency on dynamic axes)
   is_input: bool = False  # in TF1 terminology, would be a placeholder
   is_size_value: Optional[SizeValue] = None
   output_from_modules: List[_module.ModuleEntry]

--- a/pytorch_to_returnn/torch/nn/modules/operator.py
+++ b/pytorch_to_returnn/torch/nn/modules/operator.py
@@ -516,6 +516,17 @@ class Reduce(Module):
     axes = [self._get_input_axis_to_returnn(input, axis) for axis in axes]
     return {"class": "reduce", "mode": self.mode, "axes": axes, "from": self._get_input_layer_name(input)}
 
+  def make_output_tensor_from_returnn(self, inputs_flat: List[Tensor], layer: LayerBase) -> Tensor:
+    tensor = super(Reduce, self).make_output_tensor_from_returnn(inputs_flat, layer)
+    if self.mode == "max":
+      naming = Naming.get_instance()
+      assert len(inputs_flat) == 1
+      input_entry = naming.tensors[inputs_flat[0]]
+      if input_entry.is_size_value:
+        tensor_entry = naming.tensors[tensor]
+        tensor_entry.is_size_value = input_entry.is_size_value
+    return tensor
+
 
 class Length(Module):
   """

--- a/pytorch_to_returnn/torch/nn/modules/operator.py
+++ b/pytorch_to_returnn/torch/nn/modules/operator.py
@@ -36,11 +36,11 @@ class Range(Module):
     limit, start, delta, *_ = inputs_flat
     size = None
     if isinstance(limit, Tensor):
-      assert limit.is_defined
       limit_size = limit.returnn_naming_entry.is_size_value
       if limit_size is not None:
         size = (limit_size - int(start)) // int(delta)
       else:
+        assert limit.is_defined
         limit = limit.numpy()
     if size is None:
       size = SizeValue((int(limit) - int(start)) // int(delta))

--- a/pytorch_to_returnn/torch/nn/modules/variable.py
+++ b/pytorch_to_returnn/torch/nn/modules/variable.py
@@ -5,6 +5,7 @@ import tensorflow as tf
 from returnn.tf.layers.basic import LayerBase, VariableLayer
 from typing import Tuple, List, Optional
 from ...tensor import Tensor, dtype as _DType
+from ..._C import SizeValue
 from ..parameter import Parameter
 from .module import Module
 from ....naming import Naming
@@ -138,6 +139,8 @@ class FullStatic(Module):
     entry.returnn_axis_from_torch_axis = {i: i for i in range(tensor.ndim)}
     for dim in range(tensor.ndim):
       tensor.shape[dim].dim_tag = entry.returnn_data.dim_tags[dim]
+    if isinstance(self.fill_value, SizeValue):
+      entry.is_size_value = self.fill_value
     return tensor
 
 

--- a/pytorch_to_returnn/torch/nn/modules/variable.py
+++ b/pytorch_to_returnn/torch/nn/modules/variable.py
@@ -133,9 +133,11 @@ class FullStatic(Module):
     from ..._C import from_numpy
     tensor = from_numpy(numpy.full(size, self.fill_value, dtype=self.dtype.name))
     entry = naming.register_tensor(tensor)
-    entry.is_const = True
+    entry.is_const = not any(isinstance(x, Tensor) for x in inputs_flat)
     entry.returnn_data = layer.output
     entry.returnn_axis_from_torch_axis = {i: i for i in range(tensor.ndim)}
+    for dim in range(tensor.ndim):
+      tensor.shape[dim].dim_tag = entry.returnn_data.dim_tags[dim]
     return tensor
 
 

--- a/tests/test_layers.py
+++ b/tests/test_layers.py
@@ -1542,6 +1542,23 @@ def test_full():
   verify_torch_and_convert_to_returnn(model_func, inputs=x)
 
 
+def test_full_out_dim_tag():
+  n_batch, n_feat = 3, 5
+
+  def model_func(wrapped_import, inputs: torch.Tensor):
+    if typing.TYPE_CHECKING or not wrapped_import:
+      import torch
+    else:
+      torch = wrapped_import("torch")
+    full = torch.full((inputs.shape[0],), 42)
+    arange = torch.arange(full.shape[0])
+    return arange + full
+
+  rnd = numpy.random.RandomState(42)
+  x = rnd.normal(0., 1., (n_batch, n_feat)).astype("float32")
+  verify_torch_and_convert_to_returnn(model_func, inputs=x)
+
+
 def test_const_with_batch_and_gather():
   n_batch, n_time = 3, 7
 

